skr/webhook/internal/handler.go

```
@@ -0,0 +1,343 @@
+	// send request to kcp
+	watcherEvent := &types.WatcherEvent{
+		KymaCr:    kymaName,
+		Namespace: watched.Namespace,
+		Name:      watched.Name,
+	}
+	postBody, err := json.Marshal(watcherEvent)
+	if err != nil {
+		h.Logger.Error(err, "")
+		return metav1.StatusFailure
+	}
+
+	responseBody := bytes.NewBuffer(postBody)
+
+	kcpIP := os.Getenv("KCP_IP")
+	kcpPort := os.Getenv("KCP_PORT")
+	contract := os.Getenv("KCP_CONTRACT")
+	component := "manifest"
+
+	if kcpIP == "" || kcpPort == "" || contract == "" {
+		return metav1.StatusSuccess
+	}
+
+	url := fmt.Sprintf("http://%s/%s/%s/%s", net.JoinHostPort(kcpIP, kcpPort),
+		contract, component, config.EventEndpoint)
+
+	h.Logger.Info("KCP", "url", url)
+	//nolint:gosec
+	resp, err := http.Post(url, "application/json", responseBody)
+	if err != nil {
+		h.Logger.Error(err, "")
+		return metav1.StatusFailure
+	}
+
+	var response interface{}
+	if err = yaml.NewYAMLOrJSONDecoder(resp.Body, defaultBufferSize).Decode(response); err != nil {
+		h.Logger.Error(err, "response from KCP could not be unmarshalled")
+		return metav1.StatusFailure
+	}
+	h.Logger.Info("response", response)
+	return metav1.StatusSuccess
+}
+
+func (h *Handler) unmarshalRawObj(writer http.ResponseWriter, rawBytes []byte, response responseInterface,
+	resourceKind string,
+) validateResource {
+	if err := json.Unmarshal(rawBytes, response); err != nil || response.isEmpty() {
+		return h.invalidAdmissionReviewObj(writer, resourceKind, err)
+	}
+	return h.validAdmissionReviewObj("", "")
+}
+
+func (h *Handler) getKymaAndResourceListFromConfigMap(ctx context.Context) (*unstructured.Unstructured, []Resource) {
+	// fetch resource mapping ConfigMap
+	configMap := v1.ConfigMap{}
+	err := h.Client.Get(ctx, client.ObjectKey{
+		Name:      "skr-webhook-resource-mapping",
+		Namespace: "default",
+	}, &configMap)
+	if err != nil {
+		h.Logger.Error(err, "could not fetch resource mapping ConfigMap")
```

if the configMap is nil here because err != nil then the next call to configMap.Data will cause a Nilpointer as configMap is nil

---

I think this also needs retry / resilience logic

---

it should throw a notfound error in that case

---

https://github.com/kyma-project/kyma-watcher/issues/42